---
title: Reanalysis of DNA methylation analyses in archived newborn bloodspots reveal
  signals associated with fetal alcohol syndrome, maternal smoking, gender, and ethnicity
author: "David McGaughey"
date: "July 7, 2016 | September 22, 2017 | January, 2018"
output:
  html_notebook:
    toc: yes
  html_document:
    df_print: paged
    toc: yes
    keep_md: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Note

In recognition of the rapid pace of analytical tools and approaches to EWAS and Illumina 450k arrays, we conducted an analysis of our dataset guided by the recommendations of Maksimovic et al. (http://f1000research.com/articles/5-1281/v3). This Rmarkdown document contains the full code used to generate the plots and tables presented. 

```{r}
# package loading
library(data.table)
library(tidyverse)
library(annotatr)
library(ggthemes)
library(ggrepel)
library(ggsci)
library(cowplot)
library(limma)
library(minfi)
library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
library(IlluminaHumanMethylation450kmanifest)
library(GenomicFeatures)
library(RColorBrewer)
library(missMethyl)
library(matrixStats)
library(Gviz)
library(DMRcate)
library(stringr)
library(data.table)


sessionInfo()
```

## Pull in annotation, metadata, and raw Illumina 450k data from the idat files

```{r}
# pull annotation
ann450k <-  getAnnotation(IlluminaHumanMethylation450kanno.ilmn12.hg19)

# bring in raw data (Illumina .idat)
# data and code available at:
#   https://github.com/davemcg/fetal_alcohol
#   GSE65428
metadata <- fread('~/git/fetal_alcohol/sentrix_id_conversion_table_demographics_v6.txt')
targets <- read.metharray.sheet("/Volumes/ThunderBay/PROJECTS/brody/fetal_alcohol/Tsai_Sample_083/",pattern="Sheet.csv$")
targets <- read.metharray.sheet("~/Desktop/fetal_alcohol/Tsai_Sample_083/",pattern="Sheet.csv$")
rgSet <- read.metharray.exp(targets=targets)
sampleNames(rgSet) <- rgSet$Sample_Name

# add sample group info to targets, fewer rows in metadata, already parsed down to remove outliers
targets <- left_join(data.frame(metadata),targets,by=c("Sample"="Sample_Name"))
targets$Sample_Group <- targets$Case.Control
targets$Sample_Name <- targets$Sample

# remove samples based on previous outlier analysis / ethnicity (see Methods of "DNA methylation analyses in archived newborn bloodspots reveal signals associated with fetal alcohol syndrome, maternal smoking, gender, and ethnicity")
keep <-  rgSet$Sample_Name %in% targets$Sample_Name
rgSet <- rgSet[,keep]

# calculate the detection p-values for poorly performing 450k probes
detP <- detectionP(rgSet)
head(detP %>% DT::datatable())

```

## Check samples for high amounts of poorly performing sites 

Generally we want well under 5% of sites failing. Error rates for our data are very low.

```{r, fig.height=2}
# plot detection p-values
# everything looks fine in this data-set
# first match up targets to detP
targets <- targets[match(colnames(detP), targets$Sample),]

pal <- brewer.pal(8,"Dark2")

barplot(colMeans(detP), col=pal[factor(targets$Sample_Group)], las=2,
        cex.names=0.8, ylab="Mean detection p-values")
abline(h=0.05,col="red")
legend("topleft", legend=levels(factor(targets$Sample_Group)), fill=pal,
       bg="white")

barplot(colMeans(detP), col=pal[factor(targets$Sample_Group)], las=2,
        cex.names=0.8, ylim=c(0,0.002), ylab="Mean detection p-values")
abline(h=0.05,col="red")
legend("topleft", legend=levels(factor(targets$Sample_Group)), fill=pal,
       bg="white")

```


## Normalize data with preprocessQuantile and plot differences in global methylation patterning

```{r, fig.width=2}
# normalize the data; this results in a GenomicRatioSet object
mSetSq <- preprocessQuantile(rgSet)

# create a MethylSet object from the raw data for plotting
mSetRaw <- preprocessRaw(rgSet)
# visualise what the data looks like before and after normalization
targets <- targets[match(mSetSq$Sample_Name, targets$Sample),]
par(mfrow=c(1,2))
densityPlot(rgSet, sampGroups=targets$Sample_Group,main="Raw", legend=FALSE)
densityPlot(getBeta(mSetSq), sampGroups=targets$Sample_Group,
            main="Normalized", legend=FALSE)
```

## Check for systematic differences in methylation patterning

MDS plotting reveals two distinct groups, which are revealed by labeling for different factors, to be gender

```{r}

# MDS plots to look at largest sources of variation
# first, line up targets and mSetSq
targets <- targets[match(mSetSq$Sample_Name, targets$Sample),]

par(mfrow=c(2,2))
plotMDS(getM(mSetSq), top=1000, gene.selection="common",
        col=pal[factor(targets$Sample_Group)])
legend("top", legend=levels(factor(targets$Sample_Group)), text.col=pal,
       bg="white", cex=0.7)

plotMDS(getM(mSetSq), top=1000, gene.selection="common",
        col=pal[factor(targets$Gender)])
legend("top", legend=levels(factor(targets$Gender)), text.col=pal,
       bg="white", cex=0.7)

plotMDS(getM(mSetSq), top=1000, gene.selection="common",
        col=pal[factor(targets$Ethnicity)])
legend("top", legend=levels(factor(targets$Ethnicity)), text.col=pal,
       bg="white", cex=0.7)

plotMDS(getM(mSetSq), top=1000, gene.selection="common",
        col=pal[factor(targets$Sample_Plate)])
legend("top", legend=levels(factor(targets$Sample_Plate)), text.col=pal,
       bg="white", cex=0.7)
```

## Remove probes that failed QC (detP), that are related to SNP variation, and have cross-reactivity/binding

```{r}

# ensure probes are in the same order in the mSetSq and detP objects
detP <- detP[match(featureNames(mSetSq),rownames(detP)),]
# remove any probes that have failed in one or more samples
keep <- rowSums(detP < 0.01) == ncol(mSetSq)
table(keep)
mSetSqFlt <- mSetSq[keep,]
mSetSqFlt

# if your data includes males and females, remove probes on the sex chromosomes
keep <- !(featureNames(mSetSqFlt) %in% ann450k$Name[ann450k$chr %in%  c("chrX","chrY")])
mSetSqFlt <- mSetSqFlt[keep,]
table(keep)

# pull in my cg sites that I'm skipping due to overlap with SNPs and 
# are cross-reactive and are on chrs X Y
cg_remove_1 <- scan('~/git/fetal_alcohol/450k_cg_sites_that_LITERALLY_are_on_SNPS.maf005_somepop_plus1bp.txt',what='character')
cg_remove_2 <- scan('~/git/fetal_alcohol/450k_cg_sites_that_overlap_dbsnp138.maf005_somepop_plus1bp.txt',what='character')
xReactive <- scan('~/git/fetal_alcohol/illumina450k_positions_to_exclude.not_including_dbsnp_overlapping.withChen.dat',what='character')
cg_remove <- c(cg_remove_1, cg_remove_2, xReactive)
cg_remove <- unique(cg_remove)

keep <- !(featureNames(mSetSqFlt) %in% cg_remove)
table(keep)
mSetSqFlt <- mSetSqFlt[keep,]           
```

## Replot MDS with probes removed
There are two groups, white and african-american. They are evenly mixed between case/control, so shouldn't influence statistical testing too much. 
```{r}
# MDS plots to look at largest sources of variation
# now with the filtered set
# again, ensure targets and mSetSqFlt line up
targets <- targets[match(mSetSqFlt$Sample_Name, targets$Sample),]
par(mfrow=c(2,2))
plotMDS(getM(mSetSqFlt), top=1000, gene.selection="common",
        col=pal[factor(targets$Sample_Group)])
legend("top", legend=levels(factor(targets$Sample_Group)), text.col=pal,
       bg="white", cex=0.7)

plotMDS(getM(mSetSqFlt), top=1000, gene.selection="common",
        col=pal[factor(targets$Gender)])
legend("top", legend=levels(factor(targets$Gender)), text.col=pal,
       bg="white", cex=0.7)

plotMDS(getM(mSetSqFlt), top=1000, gene.selection="common",
        col=pal[factor(targets$Ethnicity)])
legend("top", legend=levels(factor(targets$Ethnicity)), text.col=pal,
       bg="white", cex=0.7)

plotMDS(getM(mSetSqFlt), top=1000, gene.selection="common",
        col=pal[factor(targets$Sample_Plate)])
legend("top", legend=levels(factor(targets$Sample_Plate)), text.col=pal,
       bg="white", cex=0.7)
```

## Calculate M-values for statistical analysis and Beta values for plotting

```{r}

# calculate M-values for statistical analysis
mVals <- getM(mSetSqFlt)
#head(mVals[,1:5])

bVals <- getBeta(mSetSqFlt)
#head(bVals[,1:5])
```

## Use limma with a linear model and eBayes for significant testing to find differentially methylated positions

The approach used in the manuscript
```{r, fig.height=2, fig.width=4}
# explicitly match targets sample order (rows) with mVars
targets <- targets[match(colnames(mVals), targets$Sample),]


case_control <- factor(targets$Case.Control)
ethnicity <- factor(targets$Ethnicity)
gender <- factor(targets$Gender)
#education <- as.numeric(targets$Education)
cd8t <- as.numeric(targets$CD8T)
cd4t <- as.numeric(targets$CD4T)
nk <- as.numeric(targets$NK)
mono <- as.numeric(targets$Mono)
gran <- as.numeric(targets$Gran)
smoking <- factor(targets$Smoking)

design <- model.matrix(~0+case_control+gender+cd8t+cd4t+nk+mono+gran+ethnicity, data=targets)
colnames(design)<-c("Case","Control","Gender","CD8T","CD4","NK","MONO","GRAN","Hispanic","Other","White")
cmtx <- makeContrasts( "Case-Control", levels=design)
fit <- lmFit(mVals, design)
fit2 <- contrasts.fit(fit, cmtx)
fit2 <- eBayes(fit2)
summary(decideTests(fit2))
```

# Annotate
```{r}
# get the table of results for the first contrast 
ann450kSub <- ann450k[match(rownames(mVals),ann450k$Name),
                      c(1:4,12:19,24:ncol(ann450k))]
DMPs <- topTable(fit2, num=Inf, coef=1, genelist=ann450kSub)

# add MORE annotations
annots = c('hg19_cpgs', 'hg19_basicgenes', 'hg19_genes_intergenic')
annotations = build_annotations(genome = 'hg19', annotations = annots)       
DMP_extra <- DMPs %>% mutate(end=pos, start=pos) 
DMP_extra <- makeGRangesFromDataFrame(DMP_extra, keep.extra.columns = F)
DMP_extra <- annotate_regions(regions=DMP_extra, annotations = annotations, ignore.strand=T)
DMP_c <- DMP_extra %>% 
  data.frame() %>% 
  dplyr::group_by(seqnames, start, end) %>% 
  summarise(Symbol=paste(unique(na.omit(annot.symbol)), collapse=','), Type = paste(unique(na.omit(annot.type)), collapse=', '))
DMP_e <- left_join(DMPs %>% tibble::rownames_to_column('Probe'), DMP_c, by=c('chr'='seqnames','pos'='start' ))

head(DMPs, n=10)
head(DMPs_e, n=10)

# get beta values 
bVals <- getBeta(mSetSq)


```

```{r}
# custom fuction to make data avail for ggplot
grabber <- function(cg_id){
  data <- bVals[cg_id,]
  data <- t(data) %>% data.frame() %>% tibble::rownames_to_column('Sample')
  data$Group <- targets$Sample_Group 
  data <- data %>% gather(Probe, Beta, -Group, -Sample)
  data <- data %>% mutate(Group = case_when(Group == 'Case' ~ 'FAS',
                                            TRUE ~ 'Control'))
  return(data)
}

# volcano
volcano <- DMP_e %>% 
  mutate(Significant = case_when(adj.P.Val < 0.05 ~ 'FDR < 0.05',
                                 TRUE ~ 'Not Significant')) %>% 
  ggplot(aes(x = logFC, y = -log10(P.Value))) +
  scale_color_manual(values = c("#BC3C29FF", "grey")) +
  geom_point(aes(colour=Significant)) +
  geom_label_repel(
    data = DMP_e %>% 
      mutate(Label= case_when(Symbol=='' ~ Probe,
                              TRUE ~ paste(Probe, Symbol, sep=' | '))) %>% 
      filter(adj.P.Val < 0.01 | adj.P.Val < 0.05 & abs(logFC) > 0.4),
    aes(label = Label),
    alpha=0.5) +
  coord_cartesian(xlim=c(-0.75,0.75)) +
  ylab('-log10(P value)')

# plot most significantly differentially methylated CpGs
top_CG_up <- DMPs %>% tibble::rownames_to_column('Probe') %>% filter(logFC > 0) %>% arrange(P.Value) %>%  dplyr::slice(1:5) %>% pull(Probe)
top_CG_down <- DMPs %>% tibble::rownames_to_column('Probe') %>% filter(logFC < 0) %>% arrange(P.Value) %>% dplyr::slice(1:5) %>% pull(Probe)
up <- grabber(top_CG_up)
down <- grabber(top_CG_down)

up_plot <- ggplot(up, aes(x=Group, y=Beta, colour=Group)) + 
  facet_wrap(~Probe, ncol=10, strip.position = 'bottom') + 
  geom_boxplot(width=0.4, outlier.shape = NA,position = position_dodge2(preserve = "total")) +
  geom_quasirandom(size=0.5, alpha=0.4) +
  theme_minimal() +
  ggsci::scale_colour_nejm() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) +
  theme(strip.text.x = element_text(angle=45)) + 
  coord_cartesian(ylim=c(0,1))

down_plot <- ggplot(down, aes(x=Group, y=Beta, colour=Group)) + 
  facet_wrap(~Probe, ncol=10, strip.position = 'bottom') + 
  geom_boxplot(width=0.4, outlier.shape = NA,position = position_dodge2(preserve = "total")) +
  geom_quasirandom(size=0.5, alpha=0.4) +
  theme_minimal() +
  ggsci::scale_colour_nejm() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) +
  theme(strip.text.x = element_text(angle=45)) + 
  coord_cartesian(ylim=c(0,1))

legend <- get_legend(up_plot + theme(legend.position="left"))
prow <- plot_grid(up_plot + theme(legend.position="none"), 
                  down_plot + theme(legend.position="none"), 
                  nrow=2, align = 'hv', labels = c('B','C'))
plot_grid(plot_grid(volcano, labels=c('A')), prow, legend, nrow=1, rel_widths = c(1,1,0.1))

```

## manhattan plot
```{r}
library(qqman)
manhattan(DMPs %>% mutate(CHR=as.numeric(gsub('chr','',chr)), BP=pos, P = P.Value))
```
## Differential methylation analysis of regions
The probe-wise test did not find much of anything, so let's try the dmrcate package to find regions of differential methylation

I plot the top hit, which looks virtually identical between case and control
```{r, fig.height=10}
myAnnotation <- cpg.annotate(object = mVals, datatype = "array", what = "M",
                             analysis.type= "differential", design = design,
                             contrasts = TRUE, cont.matrix = cmtx,
                             coef = "Case-Control", arraytype = "450K")

DMRs <- dmrcate(myAnnotation, lambda=1000, C=2)
head(DMRs$results, n=50)

# convert the regions to annotated genomic ranges
results.ranges <- extractRanges(DMRs, genome = "hg19")

# set up the grouping variables and colours
groups <- pal[1:length(unique(targets$Sample_Group))]
names(groups) <- levels(factor(targets$Sample_Group))
cols <- groups[as.character(factor(targets$Sample_Group))]
samps <- 1:nrow(targets)
# draw the plot for the top DMR
par(mfrow=c(1,1))
DMR.plot(ranges=results.ranges, dmr=1, CpGs=bVals, phen.col=cols, what = "Beta",
         arraytype = "450K", pch=16, toscale=TRUE, plotmedians=TRUE, 
         genome="hg19", samps=samps)
```

## Differential variability
Test for difference in variance between groups

Nothing looks interesting
```{r, fig.height=3}
fitvar <- varFit(mVals, design = design, coef = c(1,2))
summary(decideTests(fitvar))

DMPs_var <- topTable(fitvar, num=Inf, coef=1, genelist=ann450kSub)
head(DMPs_var, n=20)

topDV <- topVar(fitvar, coef=1)

par(mfrow=c(4,4))
sapply(rownames(topDV)[1:10], function(cpg){
  plotCpg(bVals, cpg=cpg, pheno=targets$Sample_Group,
          ylab = "Beta values")
})
```

# Comparison with *DNA methylation signature of human fetal alcohol spectrum disorder* Portales-Casamar, Lussier, Kobor et al.

## hypergeometric test
Doesn't loook too impressive (3 are sig in our dataset vs 500+ in theirs), but is more than you'd expect by chance.
```{r}
kobor_sig <- readxl::read_xlsx('13072_2016_74_MOESM2_ESM.xlsx')
kobor_probes <- kobor_sig %>% pull(Probe)

# number of Kobor probes in our dataset
overlap_num <- DMPs %>% tibble::rownames_to_column('Probe') %>% filter(Probe %in% kobor_probes) %>% nrow()
# https://stats.stackexchange.com/questions/16247/calculating-the-probability-of-gene-list-overlap-between-an-rna-seq-and-a-chip-c
# https://www.biostars.org/p/90662/
phyper(q = DMPs %>% tibble::rownames_to_column('Probe') %>% filter(Probe %in% kobor_probes, P.Value < (0.05/overlap_num)) %>% nrow(),
       m = DMPs %>% filter(P.Value < 0.05/overlap_num) %>% nrow(),
       n = (DMPs %>% nrow()) - (DMPs %>% filter(P.Value < 0.05/overlap_num) %>% nrow()),
       k = length(kobor_probes),
       lower.tail = FALSE)
```

## distribution of logFC between Kobor sig probes and not Kobor sig probes
You also see a shift in the logFC in probes that were ID'ed as FAS-related in the Kobor paper. 
```{r}
library(ggplot2)
z <- DMPs %>% 
  tibble::rownames_to_column('Probe') %>% 
  mutate(Kobor_sig = case_when(Probe %in% kobor_probes ~ 'Yes',
                               TRUE ~ 'No')) %>% 
  group_by(Kobor_sig) %>% 
  summarise(mean=mean(logFC), median=median(logFC))

t.test(z$logFC ~ z$Kobor_sig, wel)

DMPs %>% 
  tibble::rownames_to_column('Probe') %>% 
  mutate(Kobor_sig = case_when(Probe %in% kobor_probes ~ 'Yes',
                               TRUE ~ 'No')) %>% 
  ggplot(aes(x=abs(logFC), colour=Kobor_sig)) + 
  geom_density() +
  theme_bw()
```

## the successes between Kobor and ours
```{r}
DMPs %>% tibble::rownames_to_column('Probe') %>% filter(Probe %in% kobor_probes, P.Value < (0.05/overlap_num))
```

## Conclusion
We have several hundred significant probes

